app/controlplane/internal/usercontext/jwt_middleware.go

```diff
@@ -0,0 +1,231 @@
+
+type JWTOption func(*options)
+type TokenProviderOption func(*providerOption)
+
+type VerifyAudienceFunc func(*jwt.Token) bool
+
+type providerOption struct {
+	providerKey        string
+	signingMethod      jwt.SigningMethod
+	keyFunc            jwt.Keyfunc
+	claims             func() jwt.Claims
+	verifyAudienceFunc VerifyAudienceFunc
+}
+
+func WithSigningMethod(method jwt.SigningMethod) TokenProviderOption {
+	return func(o *providerOption) {
+		o.signingMethod = method
+	}
+}
+
+func WithProviderKey(providerKey string) TokenProviderOption {
+	return func(o *providerOption) {
+		o.providerKey = providerKey
+	}
+}
+
+func WithClaims(f func() jwt.Claims) TokenProviderOption {
+	return func(o *providerOption) {
+		o.claims = f
+	}
+}
+
+func WithKeyFunc(keyFunc jwt.Keyfunc) TokenProviderOption {
+	return func(o *providerOption) {
+		o.keyFunc = keyFunc
+	}
+}
+
+func WithVerifyAudienceFunc(f VerifyAudienceFunc) TokenProviderOption {
+	return func(o *providerOption) {
+		o.verifyAudienceFunc = f
+	}
+}
+
+type options struct {
+	tokenProviders []providerOption
+}
+
+func WithTokenProvider(opts ...TokenProviderOption) JWTOption {
+	op := &providerOption{}
+	for _, opt := range opts {
+		opt(op)
+	}
+	return func(o *options) {
+		o.tokenProviders = append(o.tokenProviders, *op)
+	}
+}
+
+func WithJWTTokenMiddleware(opts ...JWTOption) middleware.Middleware {
+	o := &options{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return func(handler middleware.Handler) middleware.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			if header, ok := transport.FromServerContext(ctx); ok {
+				auths := strings.SplitN(header.RequestHeader().Get(authorizationKey), " ", 2)
+				if len(auths) != 2 || !strings.EqualFold(auths[0], bearerWord) {
+					return nil, ErrMissingJwtToken
+				}
+				jwtToken := auths[1]
+				var (
+					tokenInfo *jwt.Token
+					err       error
+				)
+
+				tokenProviderLen := len(o.tokenProviders)
+				for idx, provider := range o.tokenProviders {
+					tokenInfo, err = runProviderValidator(provider, jwtToken)
+
+					// Check if it's the last provider and still failed
+					if err != nil {
+						if idx != tokenProviderLen-1 {
+							continue
+						}
+
+						return nil, err
+					}
+
+					jwtAuthContext := JWTAuthContext{
+						Claims: tokenInfo.Claims,
+					}
+
+					switch provider.providerKey {
+					case RobotAccountProviderKey:
+						jwtAuthContext.IsRobotAccount = true
+					case APITokenProviderKey:
+						jwtAuthContext.IsAPIToken = true
+					default:
+						return nil, ErrNeedTokenProvider
+					}
+
+					ctx := NewJWTAuthContext(ctx, jwtAuthContext)
+
+					//nolint:staticcheck
+					return handler(ctx, req)
+				}
+			}
+
+			return nil, ErrWrongContext
+		}
+	}
+}
+
+// runProviderValidator runs the token parser for the given provider
+func runProviderValidator(provider providerOption, jwtToken string) (*jwt.Token, error) {
+	if provider.keyFunc == nil {
```

Comment:
this is based on some other code right? I want to make sure we are not reinventing the wheel :)

Reply:
This is base on the jwtmiddleware.Serve inner logic. A few problems and why we can not reuse their code directly:
- Such middleware whenever it success, sends the request to the following middleware in the stack, making it impossible to inject custom data on the context. This change, returns a token info rather than continuing with the next middleware.
- The jwt parsing happens in a block, not in any extracted function we can include along with its validation.

Reference: https://github.com/go-kratos/kratos/blob/9e66ac2f5bcb9ab18d9b8d378c5b3233c7bb0a73/middleware/auth/jwt/jwt.go#L88

Our approach:
- Reuse parsing and validation of JWT as common as possible.
- Return a token that is valid and ready to be used, rather than continue with the next middleware.
- Use that token and the configured provider to set custom flags and information.